### DISCUSSION TITLE
CAN Testing CLI

### DISCUSSION THREAD

**Discussion by johnjacobkenny** [collin80/SavvyCAN#476] [General]
Hello @collin80 :wave:

First of all, thank you for building out this awesome tool and making it freely available for all the tinkerers out there! I have been using the GUI and it is helping me work extensively with viewing as well as replaying CAN log data files (specifically busmaster log files) on a virtual CAN bus (using can-utils). :smiley: 

I was building a CAN message parsing logic, and I am using the option "Save decoded frames". So essentially I am getting two text files, one as the output from my program, and one from SavvyCAN. I am using a diff tool on linux to compare the two, and I have caught many errors in my logic since I am new to bit wise operations. Now, however it has come to a maturity. :smile: 

Now that I am getting a lot of log files, I was thinking of a way to automate this process, and I was looking for a CLI way to replay the Busmaster Log file on the virtual can bus and to get the decoded output as a file, so that the entire process doesn't need a manual intervention. If even the replaying part only works, I can even use the python cantools library to decode the frames and then do the diff to figure out which signals are wrongly decoded (if any).

If there is anything that comes to mind, please let me know. I am trying to script out this testing as much as possible. Thanks very much! :smiley:

**Comment by collin80** [comment]
Well, I don't think anything CLI is going to directly read Busmaster log files. But, SavvyCAN can load busmaster files and it can save candump files. In theory you should able to thus use it to convert the file format and then use the standard linux CLI tool canplayer to play back the file.

**Reply by johnjacobkenny** [reply]
Thanks @collin80, I tested it and it is working! I will continue with this for now. Just to add on, is there any way to run SavvyCAN as a CLI?

**Comment by rapbando** [comment]
@collin80 Do you know how can I replay a previously captured frame via wifi to a Macchina A0 running the ESP32RET firmware?
I tried capturing the TCP packets replayed with SavyCAN but I'm looking for a python or shell script so I could change the ID and DATA bytes based on what I need. 

Regards

**Reply by collin80** [reply]
Yes, the GVRET protocol is right there in open code so it would be possible to write up something in python that conforms to the GVRET protocol and thus can send frames like savvycan does. Essentially, it opens a telnet connection to ESP32RET then sends binary commands over the link. Here is the code from SavvyCAN that writes a frame:


    buffer[0] = (char)0xF1; //start of a command over serial
    buffer[1] = 0; //command ID for sending a CANBUS frame
    buffer[2] = (char)(ID & 0xFF); //four bytes of ID LSB first
    buffer[3] = (char)(ID >> 8);
    buffer[4] = (char)(ID >> 16);
    buffer[5] = (char)(ID >> 24);
    buffer[6] = (char)((frame.bus) & 3);
    buffer[7] = (char)frame.payload().length();
    for (c = 0; c < frame.payload().length(); c++)
    {
        buffer[8 + c] = frame.payload()[c];
    }
    buffer[8 + frame.payload().length()] = 0;

    sendToSerial(buffer);

So, 0xF1 00 followed by the 32 bit ID, then the bus, then the payload length, then the actual payload bytes. The telnet session defaults to text mode though. Send 0xE7 0xE7 to put it into binary mode then you send your binary traffic to send frames. The frames come back to you in pretty much the same format. You can find that in the SavvyCAN source code too (connections/gvretserial.cpp)

**Reply by rapbando** [reply]
> Send 0xE7 0xE7 to put it into binary mode then you send your binary traffic to send frames.

This is the part I was missing! 

Best regards,

L
